File: app/src/main/java/com/bopr/android/smailer/util/db/RowSet.java
package com.bopr.android.smailer.util.db;

import android.database.Cursor;
import android.database.CursorWrapper;

import androidx.core.util.Consumer;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Extended {@link CursorWrapper}.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
@SuppressWarnings("WeakerAccess")
public abstract class RowSet<R> {

    protected final Cursor cursor;

    /**
     * Creates a cursor wrapper.
     *
     * @param cursor The underlying cursor to wrap.
     */
    public RowSet(Cursor cursor) {
        this.cursor = cursor;
    }

    protected abstract R get();

    public void forEach(Consumer<? super R> action) {
        try {
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                action.accept(get());
                cursor.moveToNext();
            }
        } finally {
            cursor.close();
        }
    }

    public R findFirst() {
        try {
            cursor.moveToFirst();
            if (!cursor.isBeforeFirst() && !cursor.isAfterLast()) {
                return get();
            }
            return null;
        } finally {
            cursor.close();
        }
    }

    public long getCount() {
        try {
            return cursor.getCount();
        } finally {
            cursor.close();
        }
    }

    public <C extends Collection<R>> C collect(final C collection) {
        forEach(new Consumer<R>() {

            @Override
            public void accept(R row) {
                collection.add(row);
            }
        });
        return collection;
    }

    public List<R> toList() {
        return collect(new ArrayList<R>());
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    protected boolean isNull(String columnName) {
        return cursor.isNull(cursor.getColumnIndex(columnName));
    }

    protected String getString(String columnName) {
        return cursor.getString(cursor.getColumnIndex(columnName));
    }

    protected boolean getBoolean(String columnName) {
        return getInt(columnName) != 0;
    }

    protected int getInt(String columnName) {
        return cursor.getInt(cursor.getColumnIndex(columnName));
    }

    protected long getLong(String columnName) {
        return cursor.getLong(cursor.getColumnIndex(columnName));
    }

    protected double getDouble(String columnName) {
        return cursor.getDouble(cursor.getColumnIndex(columnName));
    }

    public static Long forLong(Cursor cursor) {
        return new RowSet<Long>(cursor) {

            @Override
            protected Long get() {
                return cursor.getLong(0);
            }
        }.findFirst();
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/util/db/RowSet.kt
package com.bopr.android.smailer.util.db

import android.database.Cursor
import com.bopr.android.smailer.util.db.DbUtil.getDouble
import com.bopr.android.smailer.util.db.DbUtil.getInt
import com.bopr.android.smailer.util.db.DbUtil.getLong
import com.bopr.android.smailer.util.db.DbUtil.getString
import com.bopr.android.smailer.util.db.DbUtil.isNull

/**
 * [Cursor] wrapper.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
abstract class RowSet<R>(private val cursor: Cursor) {

    protected abstract fun get(): R

    fun first(): R? {
        return cursor.use { cursor ->
            cursor.moveToFirst()
            if (!cursor.isAfterLast) get() else null
        }
    }

    fun count(): Int {
        return cursor.use { it.count }
    }

    fun list(): List<R> {
        return map { it }
    }

    fun <T> map(transform: (R) -> T): List<T> {
        val list = mutableListOf<T>()
        this.forEach { list.add(transform(it)) }
        return list
    }

    private fun forEach(action: (R) -> Unit) {
        cursor.use { cursor ->
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                action(get())
                cursor.moveToNext()
            }
        }
    }

    protected fun isNull(columnName: String): Boolean {
        return cursor.isNull(columnName)
    }

    protected fun getString(columnName: String): String? {
        return cursor.getString(columnName)
    }

    protected fun getBoolean(columnName: String): Boolean? {
        return getInt(columnName)?.let { it != 0 }
    }

    protected fun getInt(columnName: String): Int? {
        return cursor.getInt(columnName)
    }

    protected fun getLong(columnName: String): Long? {
        return cursor.getLong(columnName)
    }

    protected fun getDouble(columnName: String): Double? {
        return cursor.getDouble(columnName)
    }

    companion object {

        fun forLong(cursor: Cursor): Long? {
            return object : RowSet<Long>(cursor) {

                override fun get(): Long {
                    return cursor.getLong(0)
                }

            }.first()
        }
    }

}